File: statsig-java/java/src/main/java/com/statsig/StatsigUser.java
package com.statsig;

import com.statsig.internal.JacksonUtil;
import java.util.Map;

public class StatsigUser {
  private String userID;
  private Map<String, String> customIDs;
  private String email;
  private String ip;
  private String locale;
  private String appVersion;
  private String userAgent;
  private String country;
  private Map<String, String> privateAttributes;
  private Map<String, Object> custom;

  private volatile long ref;

  /** Default constructor for Jackson deserialization. */
  public StatsigUser() {}

  private static class CleaningAction implements Runnable {
    private final long ref;

    CleaningAction(long ref) {
      this.ref = ref;
    }

    @Override
    public void run() {
      StatsigJNI.statsigUserRelease(ref);
    }
  }

  protected StatsigUser(Builder builder) {
    this.userID = builder.userID;
    this.customIDs = builder.customIDs;
    this.email = builder.email;
    this.ip = builder.ip;
    this.locale = builder.locale;
    this.appVersion = builder.appVersion;
    this.userAgent = builder.userAgent;
    this.country = builder.country;
    this.privateAttributes = builder.privateAttributes;
    this.custom = builder.custom;

    initializeRef();
    ResourceCleaner.register(this, new CleaningAction(this.ref));
  }

  private void initializeRef() {
    String customIdsJson = JacksonUtil.toJson(customIDs);
    String customJson = JacksonUtil.toJson(custom);
    String privateAttributesJson = JacksonUtil.toJson(privateAttributes);

    // Pass all arguments to the JNI binding
    this.ref =
        StatsigJNI.statsigUserCreate(
            userID,
            customIdsJson,
            email,
            ip,
            userAgent,
            country,
            locale,
            appVersion,
            customJson,
            privateAttributesJson);
  }

  // Expose a way for users to force release StatsigUser
  public void close() {
    if (ref != 0) {
      StatsigJNI.statsigUserRelease(ref);
      ref = 0;
    }
  }

  public String getUserID() {
    return userID;
  }

  public Map<String, String> getCustomIDs() {
    return customIDs;
  }

  public String getEmail() {
    return email;
  }

  public String getIp() {
    return ip;
  }

  public String getLocale() {
    return locale;
  }

  public String getAppVersion() {
    return appVersion;
  }

  public String getUserAgent() {
    return userAgent;
  }

  public String getCountry() {
    return country;
  }

  public Map<String, String> getPrivateAttributes() {
    return privateAttributes;
  }

  public Map<String, Object> getCustom() {
    return custom;
  }

  long getRef() {
    return ref;
  }

  // Builder Class
  public static class Builder {
    public String userID;
    public Map<String, String> customIDs;
    public String email;
    public String ip;
    public String locale;
    public String appVersion;
    public String userAgent;
    public String country;
    public Map<String, String> privateAttributes;
    public Map<String, Object> custom;

    public Builder setUserID(String userID) {
      this.userID = userID;
      return this;
    }

    public Builder setCustomIDs(Map<String, String> customIDs) {
      this.customIDs = customIDs;
      return this;
    }

    public Builder setEmail(String email) {
      this.email = email;
      return this;
    }

    public Builder setIp(String ip) {
      this.ip = ip;
      return this;
    }

    public Builder setLocale(String locale) {
      this.locale = locale;
      return this;
    }

    public Builder setAppVersion(String appVersion) {
      this.appVersion = appVersion;
      return this;
    }

    public Builder setUserAgent(String userAgent) {
      this.userAgent = userAgent;
      return this;
    }

    public Builder setCountry(String country) {
      this.country = country;
      return this;
    }

    public Builder setPrivateAttributes(Map<String, String> privateAttributes) {
      this.privateAttributes = privateAttributes;
      return this;
    }

    public Builder setCustom(Map<String, Object> custom) {
      this.custom = custom;
      return this;
    }

    public StatsigUser build() {
      return new StatsigUser(this);
    }
  }
}

File: statsig-dotnet/src/Statsig/StatsigUser.cs
using System;
using System.Collections.Generic;
using System.Text;
using Newtonsoft.Json;

namespace Statsig
{
    public interface IStatsigUser
    {
        string? UserID { get; }
        string? Email { get; }
        string? IP { get; }
        string? UserAgent { get; }
        string? Country { get; }
        string? Locale { get; }
        string? AppVersion { get; }
        IReadOnlyDictionary<string, string>? CustomIDs { get; }
        IReadOnlyDictionary<string, object>? CustomProperties { get; }
        IReadOnlyDictionary<string, object>? PrivateAttributes { get; }
        void Dispose();
        internal ulong Reference { get; }
    }

    public class StatsigUser : IDisposable, IStatsigUser
    {
        public string? UserID { get; }
        public string? Email { get; }
        public string? IP { get; }
        public string? UserAgent { get; }
        public string? Country { get; }
        public string? Locale { get; }
        public string? AppVersion { get; }
        public IReadOnlyDictionary<string, string>? CustomIDs { get; }
        public IReadOnlyDictionary<string, object>? CustomProperties { get; }
        public IReadOnlyDictionary<string, object>? PrivateAttributes { get; }

        private readonly ulong _ref;

        internal ulong Reference => _ref;
        ulong IStatsigUser.Reference => _ref;

        public StatsigUser(StatsigUserBuilder builder)
        {
            // Assign public properties from builder so they are accessible to callers
            UserID = builder.userID;
            Email = builder.email;
            IP = builder.ip;
            UserAgent = builder.userAgent;
            Country = builder.country;
            Locale = builder.locale;
            AppVersion = builder.appVersion;
            CustomIDs = builder.customIDs;
            CustomProperties = builder.customProperties;
            PrivateAttributes = builder.privateAttributes;

            var userIdBytes = builder.userID != null ? Encoding.UTF8.GetBytes(builder.userID) : Array.Empty<byte>();
            var emailBytes = builder.email != null ? Encoding.UTF8.GetBytes(builder.email) : Array.Empty<byte>();
            var ipBytes = builder.ip != null ? Encoding.UTF8.GetBytes(builder.ip) : Array.Empty<byte>();
            var userAgentBytes = builder.userAgent != null ? Encoding.UTF8.GetBytes(builder.userAgent) : Array.Empty<byte>();
            var countryBytes = builder.country != null ? Encoding.UTF8.GetBytes(builder.country) : Array.Empty<byte>();
            var localeBytes = builder.locale != null ? Encoding.UTF8.GetBytes(builder.locale) : Array.Empty<byte>();
            var appVersionBytes = builder.appVersion != null ? Encoding.UTF8.GetBytes(builder.appVersion) : Array.Empty<byte>();
            var customIDsJson = builder.customIDs != null ? JsonConvert.SerializeObject(builder.customIDs) : null;
            var customPropertiesJson = builder.customProperties != null ? JsonConvert.SerializeObject(builder.customProperties) : null;
            var privateAttributesJson = builder.privateAttributes != null ? JsonConvert.SerializeObject(builder.privateAttributes) : null;

            unsafe
            {
                fixed (byte* userIdPtr = userIdBytes)
                fixed (byte* emailPtr = emailBytes)
                fixed (byte* ipPtr = ipBytes)
                fixed (byte* userAgentPtr = userAgentBytes)
                fixed (byte* countryPtr = countryBytes)
                fixed (byte* localePtr = localeBytes)
                fixed (byte* appVersionPtr = appVersionBytes)
                fixed (byte* customIDsPtr = customIDsJson != null ? Encoding.UTF8.GetBytes(customIDsJson) : null)
                fixed (byte* customPropertiesPtr = customPropertiesJson != null ? Encoding.UTF8.GetBytes(customPropertiesJson) : null)
                fixed (byte* privateAttributesPtr = privateAttributesJson != null ? Encoding.UTF8.GetBytes(privateAttributesJson) : null)
                {
                    _ref = StatsigFFI.statsig_user_create(userIdPtr, customIDsPtr, emailPtr, ipPtr, userAgentPtr, countryPtr, localePtr, appVersionPtr, customPropertiesPtr, privateAttributesPtr);
                }
            }
        }

        ~StatsigUser()
        {
            Dispose(false);
        }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            StatsigFFI.statsig_user_release(_ref);
        }
    }

    public class StatsigUserBuilder
    {
        internal Dictionary<string, object>? customProperties;
        internal Dictionary<string, object>? privateAttributes;
        internal Dictionary<string, string>? customIDs;
        internal string? userID;
        internal string? email;
        internal string? ip;
        internal string? userAgent;
        internal string? country;
        internal string? locale;
        internal string? appVersion;

        public StatsigUserBuilder()
        {
        }

        public StatsigUserBuilder SetUserID(string userId)
        {
            this.userID = userId;
            return this;
        }
        public StatsigUserBuilder SetEmail(string email)
        {
            this.email = email;
            return this;
        }

        public StatsigUserBuilder SetIP(string ip)
        {
            this.ip = ip;
            return this;
        }
        public StatsigUserBuilder SetUserAgent(string userAgent)
        {
            this.userAgent = userAgent;
            return this;
        }

        public StatsigUserBuilder SetCountry(string country)
        {
            this.country = country;
            return this;
        }

        public StatsigUserBuilder SetLocale(string locale)
        {
            this.locale = locale;
            return this;
        }

        public StatsigUserBuilder SetAppVersion(string appVersion)
        {
            this.appVersion = appVersion;
            return this;
        }

        public StatsigUserBuilder SetCustomIDs(Dictionary<string, string> customIDs)
        {
            this.customIDs = customIDs;
            return this;
        }

        public StatsigUserBuilder AddCustomID(string key, string value)
        {
            customIDs ??= new Dictionary<string, string>();
            customIDs[key] = value;
            return this;
        }

        public StatsigUserBuilder SetCustomProperties(Dictionary<string, object> customProperties)
        {
            this.customProperties = customProperties;
            return this;
        }

        public StatsigUserBuilder AddCustomProperty(string key, object value)
        {
            customProperties ??= new Dictionary<string, object>();
            customProperties[key] = value;
            return this;
        }

        public StatsigUserBuilder SetPrivateAttributes(Dictionary<string, object> privateAttributes)
        {
            this.privateAttributes = privateAttributes;
            return this;
        }

        public StatsigUserBuilder AddPrivateAttribute(string key, object value)
        {
            privateAttributes ??= new Dictionary<string, object>();
            privateAttributes[key] = value;
            return this;
        }

        public StatsigUser Build()
        {
            return new StatsigUser(this);
        }
    }
}